src/components/views/rooms/RoomTileIcon.tsx

```diff
@@ -101,43 +103,43 @@ export default class RoomTileIcon extends React.Component<IProps, IState> {
     };
 
     private getPresenceIcon(): Icon {
-        let newIcon = Icon.None;
+        if (!this.dmUser) return Icon.None;
+
+        let icon = Icon.None;
 
         const isOnline = this.dmUser.currentlyActive || this.dmUser.presence === 'online';
         if (isOnline) {
-            newIcon = Icon.PresenceOnline;
+            icon = Icon.PresenceOnline;
         } else if (this.dmUser.presence === 'offline') {
-            newIcon = Icon.PresenceOffline;
+            icon = Icon.PresenceOffline;
         } else if (this.dmUser.presence === 'unavailable') {
-            newIcon = Icon.PresenceAway;
+            icon = Icon.PresenceAway;
         }
 
-        return newIcon;
+        return icon;
     }
 
-    private getIcon(): Icon {
-        let defaultIcon = Icon.None;
-        this.unsubscribePresence();
+    private calculateIcon(): Icon {
+        let icon = Icon.None;
+
         if (this.props.tag === DefaultTagID.DM && this.props.room.getJoinedMemberCount() === 2) {
             // Track presence, if available
             if (isPresenceEnabled()) {
                 const otherUserId = DMRoomMap.shared().getUserIdForRoomId(this.props.room.roomId);
                 if (otherUserId) {
                     this.dmUser = MatrixClientPeg.get().getUser(otherUserId);
-                    if (this.dmUser) {
-                        this.dmUser.on('User.currentlyActive', this.onPresenceUpdate);
-                        this.dmUser.on('User.presence', this.onPresenceUpdate);
-                        defaultIcon = this.getPresenceIcon();
-                    }
+                    icon = this.getPresenceIcon();
                 }
             }
         } else {
             // Track publicity
-            defaultIcon = this.isPublicRoom ? Icon.Globe : Icon.None;
-            this.props.room.on('Room.timeline', this.onRoomTimeline);
-            this.isWatchingTimeline = true;
+            icon = this.isPublicRoom ? Icon.Globe : Icon.None;
+            if (!this.isWatchingTimeline) {
+                this.props.room.on('Room.timeline', this.onRoomTimeline);
```

Comment:
this still feels nasty :( but ok

Reply:
it's at least 1% less nasty, which is really the best I can do to avoid code duplication.

Reply:
Fair, just that this would be much shorter, something like

```
useEventEmitter(this.props.room, "Room.timeline", this.onRoomTimeline)
```

and it'd handle the `on` and `off` for you magically

Reply:
argh, wrong comment.